scriptworker/cot/verify.py

```diff
@@ -0,0 +1,1141 @@
+        None: but cleans up the handler afterwards.
+    """
+    parent_path = os.path.join(context.config['artifact_dir'], 'public', 'cot')
+    makedirs(parent_path)
+    log_path = os.path.join(parent_path, 'audit.log')
+    audit_handler = logging.FileHandler(log_path, encoding='utf-8')
+    audit_handler.setLevel(logging.DEBUG)
+    audit_handler.setFormatter(
+        logging.Formatter(fmt='%(asctime)s %(levelname)8s - %(message)s')
+    )
+    log.addHandler(audit_handler)
+    yield
+    log.removeHandler(audit_handler)
+
+
+# guess_worker_impl {{{1
+def guess_worker_impl(link):
+    """Given a task, determine which worker implementation (e.g., docker-worker) it was run on.
+
+    Currently there are no task markers for generic-worker and
+    taskcluster-worker hasn't been rolled out.  Those need to be populated here
+    once they're ready.
+
+    * docker-worker: `task.payload.image` is not None
+    * check for scopes beginning with the worker type name.
+
+    Args:
+        link (LinkOfTrust or ChainOfTrust): the link to check.
+
+    Returns:
+        str: the worker type.
+
+    Raises:
+        CoTError: on inability to determine the worker implementation
+    """
+    worker_impls = []
+    task = link.task
+    name = link.name
+    errors = []
+
+    if task['payload'].get("image"):
+        worker_impls.append("docker-worker")
+    if task['provisionerId'] in link.context.config['scriptworker_provisioners']:
+        worker_impls.append("scriptworker")
+    if task['workerType'] in link.context.config['scriptworker_worker_types']:
+        worker_impls.append("scriptworker")
+
+    for scope in task['scopes']:
+        if scope.startswith("docker-worker:"):
+            worker_impls.append("docker-worker")
+
+    if not worker_impls:
+        errors.append("guess_worker_impl: can't find a worker_impl for {}!\n{}".format(name, task))
+    if len(set(worker_impls)) > 1:
+        errors.append("guess_worker_impl: too many matches for {}: {}!\n{}".format(name, set(worker_impls), task))
+    raise_on_errors(errors)
+    return worker_impls[0]
+
+
+# get_valid_worker_impls {{{1
+def get_valid_worker_impls():
+    """Get the valid worker_impls, e.g. docker-worker.
+
+    No longer a constant, due to code ordering issues.
+
+    Returns:
+        frozendict: maps the valid worker_impls (e.g., docker-worker) to their
+            validation functions.
+    """
+    # TODO support the rest of the worker impls... generic worker, taskcluster worker
+    return frozendict({
+        'docker-worker': verify_docker_worker_task,
+        'scriptworker': verify_scriptworker_task,
+    })
+
+
+# guess_task_type {{{1
+def guess_task_type(name):
+    """Guess the task type of the task.
+
+    Args:
+        name (str): the name of the task.
+
+    Returns:
+        str: the task_type.
+
+    Raises:
+        CoTError: on invalid task_type.
+    """
+    parts = name.split(':')
+    task_type = parts[-1]
+    if task_type.startswith('build'):
```

Comment:
I put in 'startswith' because I had some logic before that allowed for build0, build1, build2...
I don't have that anymore so I can take that out.  We may need something like that again.